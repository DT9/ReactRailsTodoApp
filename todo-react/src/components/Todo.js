import React, { Component } from 'react'

class Todo extends Component {
  handleClick = () => {
    this.props.onClick(this.props.todo.id)
  }
  handleDelete = () => {
    this.props.onDelete(this.props.todo.id)
  }
  render() {
    return (
      <div className="tile">
        <span className="deleteButton" onClick={this.handleDelete}>
          X
        </span>
        <h4 onClick={this.handleClick}>
          {this.props.todo.title}
        </h4>
        <p onClick={this.handleClick}>
          {this.props.todo.body}
        </p>
      </div>
    )
  }
}

export default Todo