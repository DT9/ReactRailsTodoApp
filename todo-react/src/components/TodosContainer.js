import React, { Component } from 'react'
import axios from 'axios'
import Todo from './Todo'
import update from 'immutability-helper'
import TodoForm from './TodoForm'



class TodosContainer extends Component {
  componentDidMount() {
    axios.get('http://localhost:3001/api/v1/todos.json')
      .then(response => {
        console.log(response)
        this.setState({ todos: response.data })
      })
      .catch(error => console.log(error))
  }
  constructor(props) {
    super(props)
    this.state = {
      todos: [],
      editingId: null,
      notification: ''
    }
  }
  addNewTodo = () => {
    axios.post(
      'http://localhost:3001/api/v1/todos',
      {
        todo:
          {
            title: '',
            body: ''
          }
      }
    )
      .then(response => {
        console.log(response)
        const todos = update(this.state.todos, {
          $splice: [[0, 0, response.data]]
        })
        this.setState({ todos: todos, editingId: response.data.id })
      })
      .catch(error => console.log(error))
  }
  updateTodo = (todo) => {
    const todoIndex = this.state.todos.findIndex(x => x.id === todo.id)
    const todos = update(this.state.todos, {
      [todoIndex]: { $set: todo }
    })
    this.setState({ todos: todos, notification: 'All Changes Saved.' })
  }
  resetNotification = () => {
    this.setState({ notification: '' })
  }
  enableEditing = (id) => {
    this.setState({ editingId: id },
      () => { this.title.focus() })
  }
  deleteTodo = (id) => {
    axios.delete(`http://localhost:3001/api/v1/todos/${id}`)
      .then(response => {
        const todoIndex = this.state.todos.findIndex(x => x.id === id)
        const todos = update(this.state.todos, { $splice: [[todoIndex, 1]] })
        this.setState({ todos: todos })
      })
      .catch(error => console.log(error))
  }
  render() {
    return (
      <div>
        <button className="addTodoButton" onClick={this.addNewTodo}>
          Add Todo
        </button>
        <span className="notification">
          {this.state.notification}
        </span>
        <div>
          {this.state.todos.map((todo) => {
            if (this.state.editingId === todo.id) {
              return (<TodoForm todo={todo} key={todo.id} updateTodo={this.updateTodo} resetNotification={this.resetNotification} titleRef={input => this.title = input} />)
            }
            return (<Todo todo={todo} key={todo.id} onClick={this.enableEditing} onDelete={this.deleteTodo} />)
          })}
        </div>
      </div>
    );
  }
}

export default TodosContainer
