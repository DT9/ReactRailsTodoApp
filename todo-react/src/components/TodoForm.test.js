import React from 'react';
import ReactDOM from 'react-dom';
import Todo from './Todo';
import TodoForm from './TodoForm'
import TodosContainer from './TodosContainer'
import {shallow} from 'enzyme';
import renderer from 'react-test-renderer';
import util from 'util'
describe('TodoForm', () => {

    it('Constructor', () => {
        const div = document.createElement('div');
        const todo = { 'id': 1, 'title': 'title', 'body': 'body' }
        ReactDOM.render(<TodoForm todo={todo} key={todo.id} onClick={TodosContainer.enableEditing} onDelete={TodosContainer.deleteTodo} />, div);
        ReactDOM.unmountComponentAtNode(div);
    })
    test('Todo changes when edited', () => {
        const todo = { 'id': 1, 'title': 'title', 'body': 'body' }
        const component = renderer.create(
            <TodoForm todo={todo} key={todo.id} onClick={TodosContainer.enableEditing} onDelete={TodosContainer.deleteTodo} />,
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
        
        // manually trigger the callback
        tree.children[0].children[0].props.value = 'title1'
        tree.children[0].props.onBlur();
        // re-rendering
        tree = component.toJSON();
        expect(tree).toMatchSnapshot();
                
    });


})
