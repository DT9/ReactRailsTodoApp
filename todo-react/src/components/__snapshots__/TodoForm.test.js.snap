// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoForm Todo changes when edited 1`] = `
<div
  className="tile"
>
  <form
    onBlur={[Function]}
  >
    <input
      className="input"
      name="title"
      onChange={[Function]}
      placeholder="Enter a Title"
      type="text"
      value="title"
    />
    <textarea
      className="input"
      name="body"
      onChange={[Function]}
      placeholder="Describe your todo"
      value="body"
    />
  </form>
</div>
`;

exports[`TodoForm Todo changes when edited 2`] = `
<div
  className="tile"
>
  <form
    onBlur={[Function]}
  >
    <input
      className="input"
      name="title"
      onChange={[Function]}
      placeholder="Enter a Title"
      type="text"
      value="title"
    />
    <textarea
      className="input"
      name="body"
      onChange={[Function]}
      placeholder="Describe your todo"
      value="body"
    />
  </form>
</div>
`;
